homeassistant/components/azure_event_hub/__init__.py

```diff
@@ -15,75 +18,200 @@
+            ),
+            "eventhub_name": config[CONF_EVENT_HUB_INSTANCE_NAME],
+        }
+        conn_str_client = False
+
+    instance = hass.data[DOMAIN] = AzureEventHub(
+        hass,
+        client_args,
+        conn_str_client,
+        config[CONF_FILTER],
+        config[CONF_SEND_INTERVAL],
+        config[CONF_MAX_DELAY],
     )
-    entities_filter = config[CONF_FILTER]
 
-    client = EventHubClientAsync(
-        event_hub_address,
-        debug=True,
-        username=config[CONF_EVENT_HUB_SAS_POLICY],
-        password=config[CONF_EVENT_HUB_SAS_KEY],
-    )
-    async_sender = client.add_async_sender()
-    await client.run_async()
+    instance.initialize()
+    return await instance.async_ready
 
     encoder = JSONEncoder()
 
-    async def async_send_to_event_hub(event: Event):
+
+class AzureEventHub:
+    """A event handler class for Azure Event Hub."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        client_args: Dict[str, Any],
+        conn_str_client: bool,
+        entities_filter: vol.Schema,
+        send_interval: int,
+        max_delay: int,
+    ):
+        """Initialize the listener."""
+        self.hass = hass
+        self.queue = asyncio.Queue()
+        self._client_args = client_args
+        self._conn_str_client = conn_str_client
+        self._entities_filter = entities_filter
+        self._send_interval = send_interval
+        self._max_delay = max_delay + send_interval
+        self.async_ready = asyncio.Future()
+        self._remove_listener = None
+        self.shutdown = False
+
+    def initialize(self):
+        """Initialize the recorder, suppress logging and register the callbacks and do the first send."""
+        # suppress the INFO and below logging on the underlying packages, they are very verbose, even at INFO
+        logging.getLogger("uamqp").setLevel(logging.WARNING)
+        logging.getLogger("azure.eventhub").setLevel(logging.WARNING)
+
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.async_shutdown)
+        self._remove_listener = self.hass.bus.async_listen(MATCH_ALL, self.async_listen)
+        self.hass.async_run_job(self.async_send, None)
+        self.async_ready.set_result(True)
+
+    @callback
+    def async_shutdown(self, _: Event):
+        """Shut down the AEH by queueing None and calling send."""
+        self._remove_listener()
+        self.hass.async_create_task(self.queue.put((time.monotonic(), None)))
+        self.hass.async_create_task(self.async_send(None))
+
+    @callback
+    def async_listen(self, event: Event):
+        """Listen for new messages on the bus and queue them for AEH."""
+        self.hass.async_create_task(self.queue.put((time.monotonic(), event)))
+
+    async def async_send(self, _):
+        """Write preprocessed events to eventhub, with retry."""
+        client = self._get_client()
+        async with client:
+            while not self.queue.empty():
```

Comment:
Why do we need to exit this task and close the client on an interval? Could we instead keep this loop running and await the queue? Only cancel the task when we stop home assistant.

Reply:
I tried it that way, but it caused some issues, so seems to be a bug in the SDK with that and therefore this way is safer and works well, even when setting the interval to a much longer time.

Reply:
Ok.